src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+  if (waited != pid || !WIFSTOPPED(status) || WSTOPSIG(status) != SIGTRAP) {
+    FATAL("Failed to wait for sigtrap after PTRACE_CONT.");
+  }
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void*> DlopenInTracee(pid_t pid, std::filesystem::path path,
+                                                   uint32_t flag) {
+  // Figure out address of dlopen in libc.
+  OUTCOME_TRY(address_dlopen, FindFunctionAddress(pid, "dlopen", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee. This is used for the code and the path name.
+  const uint64_t path_length = path.string().length() + 1;  // Include terminating zero.
+  const uint64_t memory_size = kScratchPadSize + path_length;
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, memory_size));
+
+  // Write the name of the .so into memory at address_code with offset of kScratchPadSize.
+  std::vector<uint8_t> path_as_vector(path_length);
+  memcpy(path_as_vector.data(), path.c_str(), path_length);
+  const uint64_t address_so_path = address_code + kScratchPadSize;
+  auto result_write_path = WriteTraceesMemory(pid, address_so_path, path_as_vector);
+  if (result_write_path.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_path.error();
+  }
+  // We want to do the following in the tracee:
+  // return_value = dlopen(path, RTLD_LAZY);
+  // The calling convention is to put the parameters in registers rdi and rsi.
+  // So the address of the file path goes to rdi. The flag argument goes into rsi (RTLD_LAZY is
+  // equal to 1). Then we load the address of dlopen into rax and do the call. Assembly in Intel
+  // syntax (destination first), machine code on the right:
+
+  // movabs rax, address_so_path      48 b8 address_so_path
+  // mov rbx, 1                       48 c7 c3 01 00 00 00
+  // mov rdi, rax                     48 8b f8
+  // mov rsi, rbx                     48 8b f3
+  // movabs rax, address_dlopen       48 b8 address_dlopen
+  // call rax                         ff d0
+  // int3                             cc
+  MachineCode code;
+  code.AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_so_path)
+      .AppendBytes({0x48, 0xc7, 0xc3})
+      .AppendImmediate32(flag)
+      .AppendBytes({0x48, 0x8b, 0xf8})
+      .AppendBytes({0x48, 0x8b, 0xf3})
+      .AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_dlopen)
+      .AppendBytes({0xff, 0xd0})
+      .AppendBytes({0xcc});
+  auto result_write_code = WriteTraceesMemory(pid, address_code, code.GetResultAsVector());
+  if (result_write_code.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_code.error();
+  }
+
+  ExecuteOrDie(pid, original_registers, address_code);
+
+  void* return_value = absl::bit_cast<void*>(GetReturnValueOrDie(pid));
+
+  // Cleanup memory and registers.
```

Comment:
```suggestion
  // Clean up memory and registers.
```

Reply:
Are you going to address this?

Reply:
Sorry, missed this and some others. It seems not all comments show up under the "Conversation" tab. I only see the second one there. That's a bit of a mess.

Reply:
done